src/journeys_utils.js

```diff
@@ -438,14 +439,94 @@ journeys_utils.animateBannerEntrance = function(banner) {
 		}
 		else if (journeys_utils.position === 'bottom') {
-			banner.style.bottom = '0';
+			if(!journeys_utils._isSafeAreaRequired(journeys_utils.journeyLinkData)) {
+				banner.style.bottom = '0';
+			} else {
+				journeys_utils._dynamicallyRepositionBanner(banner);
+			}
 		}
         journeys_utils.branch._publishEvent('didShowJourney', journeys_utils.journeyLinkData);
 		journeys_utils.isJourneyDisplayed = true;
 	}
 	setTimeout(onAnimationEnd, journeys_utils.animationDelay);
 }
 
+journeys_utils._isSafeAreaRequired = function(journeyLinkData) {
+	if (journeyLinkData['safeAreaRequired'] && journeys_utils._detectSafeAreaDevices()) {
+		return true;
+	}
+	return false;
+}
+
+journeys_utils._resizeListener = function () {
+	if (journeys_utils.isSafeAreaEnabled) {
+		journeys_utils._resetJourneysBannerPosition(false);
+	}
+}
+
+journeys_utils._scrollListener = function () {
+	if (journeys_utils.isSafeAreaEnabled) {
+		if (window.pageYOffset > window.innerHeight) {
+			journeys_utils._resetJourneysBannerPosition(true);
+		} else {
+			journeys_utils._resetJourneysBannerPosition(false);
+		}
+	}
+}
+
+journeys_utils._dynamicallyRepositionBanner = function() {
+	journeys_utils.isSafeAreaEnabled = true;
+	// disable Journey animation to avoid lag when repositioning the banner 
+	document.getElementById('branch-banner-iframe').style.transition = "all 0s"
+	// make sure on the first journey load the position is correct
+	journeys_utils._resetJourneysBannerPosition(false, true);
+	// resize listener for Safari in-app webview resize due to bottom/top nav bar
+	window.addEventListener("resize", journeys_utils._resizeListener);
+	// scroll listener for bottom overscrolling edge case
+	window.addEventListener("scroll", journeys_utils._scrollListener);
+}
+
+journeys_utils._resetJourneysBannerPosition = function(isPageBottomOverScrolling, checkIfPageAlreadyScrollingOnFirstLoad) {
+	var bannerIFrame = document.getElementById('branch-banner-iframe');
+	var bannerHeight = bannerIFrame.offsetHeight;
+	var bannerTopDistance = bannerIFrame.offsetTop;
+	var windowHeight = window.innerHeight;
+
+	// on first load check if the page is already scrolling
+	if(checkIfPageAlreadyScrollingOnFirstLoad) {
+		if(window.pageYOffset !== 0) {
+			bannerIFrame.style.bottom = '0';
+			return false;
+		}
+	}
+
+	if (!isPageBottomOverScrolling) {
+		// always keep banner top location equal to the height specified
+		if ((windowHeight - bannerTopDistance) != bannerHeight) {
+			bannerIFrame.style.top = "" + (windowHeight - bannerHeight) + "px";
+		}
+	} else {
+		// bottom overscrolling is usually equivalent to half the banner size 
+		bannerIFrame.style.top = (windowHeight - bannerHeight) + (bannerHeight / 2) + "px";
+	}
+}
+
+// generic method to detect which devices should call journey_utls._dynamicallyRepositionBanner()
+journeys_utils._detectSafeAreaDevices = function() {
+	var isFormFactoriOS = /iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream;
```

Comment:
@code-crusher can you please explain window.MSStream?

Reply:
Updated this to use `navigator.platform`